### PR TITLE
IRB is improved with Reline and RDoc, take 2

### PR DESCRIPTION
**Pull request by aycabta**
[My previous try](https://github.com/ruby/ruby/commit/7f273ac6d0f05208b5b228da95205e20c0e8286c) was failed and [reverted](https://github.com/ruby/ruby/commit/51cec00953ff8d7baa483d3846aa1dbdb89101aa) so I'll try again.

### CODE REVIEW

**Comment by mame** [comment]
* `[RET]` → `[UP ARROW]` raises: undefined method encode for nil:NilClass (NoMethodError)
* `[DEL]` raises: "\\xDB" from ASCII-8BIT to UTF-8 (Encoding::UndefinedConversionError)

* Pushing `[RET]` in the middle of lines causes weird behavior. ↓

```
def foo(x)
  x[CURSOR]
end
```
↓push `[RET]`
```
def foo(x)
[CURSOR]
end
```
↓push `[RET]`
```
def foo(x)
  x
[CURSOR]
end
```

* `.[RET]` does not cause SyntaxError immediately

* Pushing `[RET]` in the middle of a line causes inconsistent behavior. ↓

```
1.times { p x [CURSOR]}
```
↓push `[RET]`
```
1.times { p x 
}[CURSOR]
```
But
```
1.times { p [CURSOR]x }
```
↓push `[RET]` and the code is executed immediately
```
1.times { p
x }
#=> 1
```

* `autoload? :A[RET]` and `respond_to? :Integer[RET]` does not run the code.

* Pushing `[ALT]+[RET]` multiple times causes weird result (intentional?)

```
irb(main):001:0> ^<compiled>:1: warning: encountered \r in middle of line, treated as a mere space
irb(main):001:0> ^M^M
irb(main):001:0> 
```

* Changing the width of the terminal causes weird result (SIGWINCH seems not to be handled)

**Comment by aycabta** [comment]
In vi mode,

```
def hoge[cursor]
end
```

and press `jx`, IRB crashes.

**Comment by aycabta** [comment]
I'll re-try the CI challenge with split commits.

File: lib/irb.rb [hunk 1]
```diff
@@ -60,21 +60,22 @@
 #     -W[level=2]       Same as `ruby -W`
 #     --inspect         Use `inspect' for output (default except for bc mode)
 #     --noinspect       Don't use inspect for output
+#     --reidline        Use Reidline extension module
+#     --noreidline      Don't use Reidline extension module
```
Comment (k0kubun) [inline lib/irb.rb]: please show them in `irb --help` too

**Comment by aycabta** [comment]
Strange behavior is occurred with `:SymbolStartedWithLargeLetter`.

**Comment by k0kubun** [comment]
Looks like the merge was 260235ce871c3e7718af8d612f1a8ed400b56070 and 559dca509d2a98584b09c7d9a6d74749ce793ad7. Please use `make merge-github PR=2144` or rebase the PR before committing to trunk next time.

**Pull request closed without merging**